vowpalwabbit/learner.h

```diff
@@ -186,18 +228,44 @@ struct learner
   func_data end_pass_fd;
   func_data end_examples_fd;
   func_data finisher_fd;
-
   std::shared_ptr<void> learner_data;
+
   learner(){};  // Should only be able to construct a learner through init_learner function
+
  public:
   prediction_type_t pred_type;
   size_t weights;  // this stores the number of "weight vectors" required by the learner.
   size_t increment;
   bool is_multiline;  // Is this a single-line or multi-line reduction?
+  std::string name;   // Name of the reduction.  Used in VW_DBG to trace nested learn() and predict() calls
```

Comment:
if I have two instances of learner objects, and I fetch the name of each and both reply back with the same exact name what can one infer from that?

Reply:
This is used to generate trace of execution stack and report state of reductions in context.  No assumption is made about the uniqueness of names.

Here is an example:

    - [shared_feature_merger.learn]
        - [cb_sample.learn]
            - [cb_explore_adf-greedy.learn]
                - [cb_adf.learn]
                  - cs_ldf_learn: ex=141, offset=0, id=0
                    - [csoaa_ldf_rank.predict]
                        - [scorer-identity.predict]
                            - [gd.predict]
                            -
                        -
                        - [scorer-identity.predict]
                            - [gd.predict]
                            -
                        -
                        - [scorer-identity.predict]
                            - [gd.predict]
                            -
                        -
                    -

Reply:
I'm aware that's the only current usage of 'name' now but having a public member variable called 'name' on an object called 'learner' has implications for all the code base and for future work. 'name' has very strong connotations and I believe it will cause confusion at some point since 'name' only returns the top most level reduction name and hiding away whatever else is on that learner stack. Two learner instances might return the same name but will behave differently (when calling predict/learn) if they have different stacks but only coincide on the top level which is a bit dissonant.

Reply:
I agree with @lalo. A publicly accessible field called `name` seems dangerous here.

Reply:
"name" does make a strong statement and does not reflect the semantics.  Does trace_name or trace_reduction_name work?  If we find other uses for this in the future and provide stronger guarantees, we can change the name to reflect the semantics.  Does that address the concern?

Reply:
Isn't passing in a name at the learner for each reduction opposite of what we were discussing for this PR (https://github.com/VowpalWabbit/vowpal_wabbit/pull/2531)?

Reply:
The objection I am aware of was in exposing more metadata about the reduction in parse_args (i.e. outside the reduction).

Reply:
I think it was about not asking the reduction writer to pass in a name and having them have to think about this too, no? @lalo ?

Reply:
Yes, both of you are right. We can ask John what he thinks since he was the one that raised concerns with that in the previous discussion.